src/sagas/pushNotification.js

```diff
@@ -0,0 +1,440 @@
+  let dispatch;
+  yield put((_dispatch) => {
+    dispatch = _dispatch;
+  });
+
+  const useWalletService = yield select((state) => state.useWalletService);
+
+  // If the user is not using the wallet-service,
+  // we need to initialize the wallet on the wallet-service first
+  let walletService;
+  if (!useWalletService) {
+    // Set urls for wallet service
+    config.setWalletServiceBaseUrl(WALLET_SERVICE_MAINNET_BASE_URL);
+    config.setWalletServiceBaseWsUrl(WALLET_SERVICE_MAINNET_BASE_WS_URL);
+    const network = new Network(NETWORK);
+
+    const pin = yield call(showPinScreenForResult, dispatch);
+    walletService = new HathorWalletServiceWallet({
+      requestPassword: pin,
+      seed: walletUtil.getWalletWords(pin),
+      network,
+      enableWs: false,
+    });
+
+    try {
+      yield call(walletService.start.bind(walletService), {
+        pinCode: pin,
+        password: pin,
+      });
+    } catch (error) {
+      yield put(pushLoadWalletFailed({ error }));
+    }
+  } else {
+    walletService = yield select((state) => state.wallet);
+  }
+
+  yield put(pushLoadWalletSuccess({ walletService }));
+}
+
+/**
+ * This function is the actual opt-in of a user to the Push Notifications feature.
+ * It should be called when the push notifications are not loaded and/or registered.
+ * This should load the wallet on the wallet-service and register it with the deviceId.
+ * @param {{ payload: { deviceId: string } }} action
+ */
+export function* firstTimeRegistration({ payload: { deviceId } }) {
+  yield put(pushLoadWalletRequested());
+
+  // wait for the wallet to be loaded
+  const [loadWalletSuccess, loadWalletFail] = yield race([
+    take(types.PUSH_WALLET_LOAD_SUCCESS),
+    take(types.PUSH_WALLET_LOAD_FAILED)
+  ]);
+
+  if (loadWalletFail) {
+    yield put(pushRegisterFailed());
+    return;
+  }
+
+  const { walletService } = loadWalletSuccess.payload;
+  try {
+    const { success } = yield call(pushLib.PushNotification.registerDevice, walletService, {
+      pushProvider: Platform.OS,
+      deviceId,
+      enablePush: true,
+    });
+
+    if (success) {
+      const enabledAt = Date.now();
+      STORE.setItem(pushNotificationKey.enabledAt, enabledAt);
+      STORE.setItem(pushNotificationKey.hasBeenEnabled, true);
+      yield put(pushRegisterSuccess({ enabled: true, hasBeenEnabled: true, enabledAt }));
+    } else {
+      // NOTE: theoretically, this should never happen because when the client call fails, it throws an error
+      yield put(pushRegisterFailed());
+    }
+  } catch (error) {
+    console.error('Error registering device for the first time: ', error.cause);
+    yield put(pushRegisterFailed());
+  }
+}
+
+/**
+ * This function is responsible for registering the device on the wallet-service in the event
+ * of renewing the registration.
+ */
+export function* registration({ payload: { enabled, showAmountEnabled, deviceId } }) {
```

Comment:
Except for one line (`STORE.setItem(pushNotificationKey.hasBeenEnabled, true);`), which is ok to be in this method too, isn't this method the same as `firstTimeRegistration`? 

I feel we should have only one of them and remove the duplicated code, what do you think?

Reply:
Sure! I refactored it to remove first registration. Now we only have registration.